src/transformers/cache_utils.py

```diff
@@ -0,0 +1,89 @@
+from abc import ABC, abstractmethod
+from typing import Dict, List, Tuple, TypeVar
+import torch
+
+T = TypeVar("T")
+
+
+class Cache(ABC):
+    def __init__(self) -> None:
+        self.cache: Dict[int, Tuple[torch.Tensor]] = {}
+        self.layer_idx = 0
+
+    @abstractmethod
+    def update_pre_rotation(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
+        pass
+
+    @abstractmethod
+    def update(self, key_states, value_states) -> None:
+        pass
+
+    def __getitem__(self, index: int):
+        return self.cache[self.layer_idx][index]
+
+    def set_layer_idx(self, layer_idx: int) -> None:
+        self.layer_idx = layer_idx
+
+    def __bool__(self) -> bool:
+        return bool(self.cache) and self.layer_idx in self.cache
+
+class DynamicCache(Cache):
+    def update_pre_rotation(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
+        pass
+
+    def update(self, key_states: torch.Tensor, value_states: torch.Tensor) -> None:
+        kv_states = torch.cat([key_states[None, :], value_states[None, :]], dim=0)
```

Comment:
actually upon a second thought this is probably a bad idea to concat k,v states here because it unnecessarily allocates memory. Maybe it would be better to just keep to dictionaries `key_cache`, `value_cache`? This way we save one `torch.cat(...)` operation. Wdty @gante?

Reply:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8.